### DISCUSSION TITLE
Chinese Pinyin Search

### DISCUSSION THREAD

**Discussion by dyphire** [tomasklaen/uosc#656] [Ideas]
Edit: due to mpv cannot switch the language IME when typing  on Windows, the uosc script is limited to searching using latin character.
I have implemented a pinyin search method for Chinese on my personal branch: https://github.com/dyphire/uosc/commit/101ff2d, and Chinese users who have a need for it can try using and testing it.
PS: This will not break the existing latin character search.

**Update**: add support for searching Chinese initials.

Since this is a Chinese specific solution rather than unicode, it will not enter the main branch.
Here are the renderings:

|  input-search  | browser-search |
| ---------------- | ---------------- |
| ![image](https://github.com/tomasklaen/uosc/assets/61936050/0bb37036-ec7e-4018-ad47-59fe63d932fe)   |  ![image](https://github.com/tomasklaen/uosc/assets/61936050/8dbc5dc8-cd8b-44aa-8b7f-3e846b7c3273) |

**Comment by christoph-heinrich** [comment]
> Due to mpv not yet supporting non English input, the uosc script cannot support unicode searches for non English languages.

https://github.com/mpv-player/mpv/blob/0b705983584691c6d7b1ee19c4b6950fa5334326/input/input.c#L483-L487
German multi byte utf8 characters (`äöü`) work fine for me. If Chinese characters don't work for you then that would be a bug in mpv.

You could use already existing utf8 stuff from uosc. That loop in `pinyin()` looks like it could use `utf8_iter()`, which would not only be simpler but also much faster.
Also here is a much faster way of getting individual utf8 characters from a string. https://github.com/tomasklaen/uosc/blob/7ffd2ab705ee6cac2d3c3b9245e5f80b7e26aeca/scripts/uosc/lib/text.lua#L80-L87

And if you were to replace the character only if it is in `pyTable` instead of checking for byte length then that would open up the possibility of doing the same thing for other alphabets. Then we only need to append which characters should be replaced with which other ones to the table and any language could be searched with latin characters. Always generating all of that during a search seems terribly wasteful, so we could prepare such things at the start of a search, similar to how I did in https://github.com/christoph-heinrich/uosc/commit/bb3a79aa6eca82f2e895eeb521f74358ad31f004

**Reply by dyphire** [reply]
> https://github.com/mpv-player/mpv/blob/0b705983584691c6d7b1ee19c4b6950fa5334326/input/input.c#L483-L487 German multi byte utf8 characters (`äöü`) work fine for me. 
If Chinese characters don't work for you then that would be a bug in mpv.
> 

The problem here is that mpv cannot switch the language IME when typing on Windows, so Chinese characters cannot be input.

I have seen related issues on the mpv official issue tracker, and developers have noticed this on Windows, but it is not yet clear what caused this issue.

Your suggestion looks good. If I have time, I will try to improve it.

**Reply by christoph-heinrich** [reply]
> The problem here is that mpv cannot switch the language IME when typing on Windows, so Chinese characters cannot be input.

Ah I didn't even think of that.

@tomasklaen do we want such a latin character representation of menu items thing? It would be easy to implement as it would be mostly using code that's already lying around anyway. I'm not sure about the translation tables, they could be json files like the intl stuff.

**Reply by dyphire** [reply]
> @tomasklaen do we want such a latin character representation of menu items thing? It would be easy to implement as it would be mostly using code that's already lying around anyway. I'm not sure about the translation tables, they could be json files like the intl stuff.

This looks great, we can even use it in sorting instead of just searching.

**Reply by tomasklaen** [reply]
> do we want such a latin character representation of menu items thing?

I don't understand the question :) Can you elaborate?

**Reply by dyphire** [reply]
Based on @christoph-heinrich  suggestion, I refactored the implementation and now it has scalability. ref https://github.com/dyphire/uosc/commit/cb7a3b6

> Always generating all of that during a search seems terribly wasteful, so we could prepare such things at the start of a search, similar to how I did in https://github.com/christoph-heinrich/uosc/commit/bb3a79aa6eca82f2e895eeb521f74358ad31f004

I hope to make this improvement after your relevant PR is merged.

**Reply by christoph-heinrich** [reply]
> > do we want such a latin character representation of menu items thing?
> 
> I don't understand the question :) Can you elaborate?

Pinyin is specific to Chinese, but the same concept can also be applied to other languages e.g. [russian](https://en.wikipedia.org/wiki/Russian_Latin_alphabet). We could make a translation table to create a latinized version of title and hint and then search those too, like dyphire is already doing here for Chinese.

**Reply by christoph-heinrich** [reply]
> I hope to make this improvement after your relevant PR is merged.

You mean the fuzzy search? I'm not sure that will ever get merged, and after recent menu changes that will require a rebase that I expect to be non trivial and I can't be bothered to do that right now.

**Reply by tomasklaen** [reply]
> Pinyin is specific to Chinese, but the same concept can also be applied to other languages e.g. [russian](https://en.wikipedia.org/wiki/Russian_Latin_alphabet). We could make a translation table to create a latinized version of title and hint and then search those too, like dyphire is already doing here for Chinese.

I mean I don't understand what the actual problem is, or the proposed solution. From what I've gathered, when people input chinese characters, the `any_unicode` event gives us latin? And you want to have some tables that convert it back to chinese?

**Reply by dyphire** [reply]
> I mean I don't understand what the actual problem is, or the proposed solution. From what I've gathered, when people input chinese characters, the `any_unicode` event gives us latin? And you want to have some tables that convert it back to chinese?

The problem here is that due to the limitations of mpv, we are unable to input Chinese characters for search. 
I am trying to convert Chinese characters to latin characters to improve the current situation and achieve using Pinyin for search.

What @christoph-heinrich  wants to say is whether similar processing is needed for other non latin characters.

**Reply by christoph-heinrich** [reply]
> when people input chinese characters, the any_unicode event gives us latin?

Does it? I have no way of testing it, but from what I understood there is actually no way for people to input chinese characters because of some IME thing. Idk I've never had to deal with that, but they don't have enough keys on their keyboard to directly input their symbols, so they have this conversion thingy, and I guess that doesn't work with mpv. As a consequence they have to input latin characters, but the menu items use chinese characters, and so to be able to search for those they need to be converted into latin characters.

**Reply by tomasklaen** [reply]
This sounds like an upstream problem, and an effort to fixing it would be better spent there. I really don't want the hacks and tables necessary to make it work in uosc.

I'd like to help somehow with fixing it, but not by bloating uosc with input method handling.

Maybe something like new API messages, where scripts could send:

- `input-method <'native'|'message'>` - when script sends `input-method message` to us, the menu will not bind `any_unicode` listener anymore, and instead will only listen on:
- `character-input <char>` - a message 3rd party scripts can send to input a character

This way there could be 3rd party scripts created that convert latin inputs to chinese/etc characters, and send those to uosc.

**Reply by christoph-heinrich** [reply]
That sounds more like a work around then the latinisation thing (which is apparently called romanization[[1](https://en.wikipedia.org/wiki/Latinisation)][[2](https://en.wikipedia.org/wiki/Romanization)]) and also more complicated.

You're welcome to try and fix Chinese input on the mpv side, but I still think that romanization would be a useful feature even if input of native characters works (no need to switch keyboard layouts for people that speak e.g. english and chinese).

**Reply by dyphire** [reply]
> This sounds like an upstream problem, and an effort to fixing it would be better spent there. I really don't want the hacks and tables necessary to make it work in uosc.

mpv related issues: https://github.com/mpv-player/mpv/issues/9168
Yes, I mentioned from the beginning that this is a processing specific to Chinese characters. This is not prepared to enter the main branch.

> Maybe something like new API messages, where scripts could send:
> 
> * `input-method <'native'|'message'>` - when script sends `input-method message` to us, the menu will not bind `any_unicode` listener anymore, and instead will only listen on:
> * `character-input <char>` - a message 3rd party scripts can send to input a character
> 
> This way there could be 3rd party scripts created that convert latin inputs to chinese/etc characters, and send those to uosc.

This method cannot achieve correct character conversion. For Chinese characters, the number of characters is extremely large, exceeding 60000 in existing international standards and 80000 in China's official standards.

It is feasible to convert commonly used Chinese characters into latin characters through Pinyin, but the opposite cannot be achieved. 
Just read this file of [zh.json](https://github.com/dyphire/uosc/blob/dev/scripts/uosc/char-conv/zh.json) and you will realize how terrifying the idea of converting Latin characters into Chinese characters. This is equivalent to the 3rd party script having to implement a complete IME and interactively selectable UI on its own, and I cannot see the feasibility of this.

We either wait for the mpv upstream to fix the related issues or let Chinese users use my temporary solution, which is why I opened a discussion.

**Reply by dyphire** [reply]
> but I still think that romanization would be a useful feature even if input of native characters works (no need to switch keyboard layouts for people that speak e.g. english and chinese).

Yes, I have to admit that this idea is very attractive.

**Reply by tomasklaen** [reply]
Oh I get the problem now. Yeah my API proposal wouldn't be much useful here.

> I still think that romanization would be a useful feature even if input of native characters works

That would mean shipping uosc with romanization tables, detecting a string has romanizable characters, loading the table, romanizing the string, and doing 2x more string comparisons for each item than we're currently doing?

And how big would be the table for chinese? Is that the one in the `pinyin.lua` file by dyphire? It's almost 90kb.

Eh, I don't feel good about this.

**Reply by christoph-heinrich** [reply]
Given the release zip size it seems to me that shipping ~100kb of tables isn't really a problem anymore (I don't expect any other language to get anywhere close to chinese in size).
Search performance is pretty good, even with close to 3000 subtitles lines it's practically instant, so I don't expect twice the comparisons to be a problem either.

Loading all of that into RAM is a  bit memory intensive, but we could restrict what gets loaded to what's configured in `options.languages`. After all there isn't any point in romanizing a language the user doesn't understand.

Any other objections? :)

**Reply by dyphire** [reply]
Yes, due to ziggy expanding the release size to 5+MB, I also believe that table size is no longer an issue. 
Given the good performance and extensibility of this implementation, merging it into the main branch would be a good idea.

If there's no problem, I can open a PR for this.

**Reply by tomasklaen** [reply]
@christoph-heinrich if it won't even be `require()`d when not needed, then sure.

**Reply by christoph-heinrich** [reply]
How come `require()` is such a big deal? I thought I'd put it in`intl.lua` because it's part of multi-language stuff.
Or do you mean making them lua tables instead of json, because converting those is kinda silly.

**Reply by tomasklaen** [reply]
I mean the 100kb table to only be loaded when needed. I wasn't sure if it's going to be a json or lua.

**Reply by dyphire** [reply]
I don't understand where the difference lies here, as long as the table exists in the scripts folder, it will be loaded when mpv starts. Preloading also helps improve performance. Moreover, the current implementation within uosc is already to set the content of the on-demand read table according to the user's `options.languages`, not to load and read it for everyone. ref [char_conv.lua#L6-L14](https://github.com/dyphire/uosc/blob/f30eba50631542806745b995aece953dfaca99b8/src/uosc/lib/char_conv.lua#L6-L14)

I think what's really worth optimizing here is @christoph-heinrich  previous suggestion that we can prepare this content before searching instead of generating it during the search process.

**Comment by dyphire** [comment]
After that, I tested the support of some IME under mpv on Windows: Chinese, Japanese, Korean, Russian, German, French.
Then I found out that the IME that cannot be used when mpv is in focus is Chinese, Japanese, and Korean. Those other IME can be used normally, which means CJK characters cannot be type in mpv.

This means that probably only CJK characters need romanization support.

**Reply by christoph-heinrich** [reply]
Huh that's weird, I wonder why it doesn't work. 
If anyone wants to debug this, the function to start at would be [here](https://github.com/mpv-player/mpv/blob/d32f1aac3ffd2e7b766773446c4510a2cad397fb/input/input.c#L452-L496).
From a quick look the symbols should be within the range `MP_KEY_IS_UNICODE` checks for, so no idea why it doesn't work.

**Reply by dyphire** [reply]
This is more like a system platform specific issue, so I think what needs to be debugged in Windows is [w32_common.c](https://github.com/mpv-player/mpv/blob/master/video/out/w32_common.c). 
Someone raised a possibility there: https://github.com/mpv-player/mpv/issues/9168#issuecomment-1748963991

**Comment by debugzxcv** [comment]
如果在搜索时直接粘贴中文则无法获得结果：
![111](https://github.com/tomasklaen/uosc/assets/5125392/be18b7dd-83d2-4b1d-9f99-454bd595dd3f)

**Reply by dyphire** [reply]
fixed by https://github.com/dyphire/uosc/commit/101ff2d
